### DISCUSSION TITLE
LedStrip run command syntax

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/devicescript#551] [Q&A] [answered]
I have been having a great time with DeviceScript so far and am looking to port over some work I did in the C++ WLED library.

I am struggling to understand the LedStrip.run() command syntax though after looking at the docs and trying as many things as I could think of ([LedStrip Docs](https://microsoft.github.io/devicescript/api/clients/ledstrip)). I was wondering what some basic command syntax might look like for newing up a strip with 30 LEDS, setting all LEDs to blue,  then setting all LEDs to red, then rotating the strip back 3 pixels. Any light program examples would be helpful!

Keep up the awesome work.

**Comment by mmoskal** [comment, marked as answer]
There are some examples here https://github.com/microsoft/pxt-jacdac/blob/master/led-strip/client.ts however... 

The LedStrip "language" was invented to send the RGB LED animation updates for long strips over relatively slow Jacdac hardware link (effectively 50 kilobytes per second). We also have a simpler LED service that sends all the LED values in one packet and is limited to 64 LEDs. Either one of these can be implemented by an external Jacdac module connected over a Jacdac cable, like [LED Ring](https://microsoft.github.io/jacdac-docs/devices/forward-education/ledlightsv10/) or [LED Strip adapter](https://microsoft.github.io/jacdac-docs/devices/kittenbot/rgbstripv10/)

Now, most likely most of our current users don't care about these Jacdac modules (as they are hard to find) and instead would rather connect the LED strip directly to Pico or ESP32. I'm currently leaning towards adding a native driver on both of these that would be exposed similar to the [SPI APIs](https://microsoft.github.io/devicescript/api/spi) (not through a Jacdac service). That makes it a little more difficult for us to support simulation but is otherwise simpler than the LedStrip. The API for this would be just  `await sendRGB(...rgb buffer...)` - this API could be supported for both directly connected RGB strips and for the LED modules (but not necessarily LedStrip modules). 

Once we have this API it would be up to users (you? ;] ) to come up with a nice animation library. Before we have this generic API, you can just implement it for the LED service only (essentially `sendRGB = led.pixel.write`) and test stuff out in simulator up to 64 LEDs.

Hope this helps and thank you for your kind words!

**Reply by ghost** [reply]
Thanks for pointing me in this direction! 

I agree with you though, most LED projects and libraries for microcontrollers typically connect directly to the microcontroller (FastLED, WLED, neopixel). Some of those libraries are capable of running multi thousand LED long strands in parallel from a single ESP32 and I would love to try and help enable that to be possible with DeviceScript. I am not sure if the current limitations of DeviceScript prevent that from being possible, but want to try and help make it possible. Could I assist in any way helping create the native driver?

**Reply by pelikhan** [reply]
(for completeness, i am porting the encoder from makecode - https://github.com/microsoft/devicescript/pull/557)

**Reply by pelikhan** [reply]
@zacharyrogers1 take a look at the PR https://github.com/microsoft/devicescript/pull/580 we are working towards a LED driver.

The approach we decided has been to rely on the Led service (not the ledstrip with the funny rendering language) and allow for an arbitrary number of LEDs. We updated DeviceScript to support simulation for those.

The LED painting operation hinge on the "PixelBuffer" class which acts as a thin layer over a Buffer that will be probably DMAed to the LED strip.n The PixelBuffer class is quite barebone and there are probably several features that could be added in the future like gamma correction, max-power correction.

If you have time to take a look, that would be great!

**Reply by pelikhan** [reply]
we also have a little driver that mounts a Display on a LED (matrix). so could easily treat it as a palette screen.
